rasa/engine/model_storage.py

```diff
@@ -0,0 +1,305 @@
+            )
+
+        yield directory
+
+    def create_model_package(
+        self,
+        model_archive_path: Union[Text, Path],
+        train_schema: GraphSchema,
+        predict_schema: GraphSchema,
+        domain: Domain,
+        model_metadata: Dict[Text, Any],
+    ) -> None:
+        """Creates a model archive containing all data to load and run the model.
+
+        Args:
+            model_archive_path: The path to the archive which should be created.
+            train_schema: The schema which was used to train the graph model.
+            predict_schema: The schema for running predictions with the trained model.
+            domain: The `Domain` which was used to train the model.
+            model_metadata: Any additional metadata which should be stored with the
+                model.
+        """
+        logger.debug(f"Start to created model package for path '{model_archive_path}'.")
+
+        with tempfile.TemporaryDirectory() as temp_dir:
+            temporary_directory = Path(temp_dir)
+
+            shutil.copytree(
+                self._storage_path, temporary_directory / MODEL_ARCHIVE_COMPONENTS_DIR
+            )
+
+            self._persist_schemas(train_schema, predict_schema, temporary_directory)
+            self._persist_metadata(domain, model_metadata, temporary_directory)
+
+            with tarfile.open(model_archive_path, "w:gz") as tar:
+                tar.add(temporary_directory, arcname="")
+
+        logger.debug(f"Model package created in path '{model_archive_path}'.")
+
+    @staticmethod
+    def _persist_schemas(
+        train_schema: GraphSchema,
+        predict_schema: GraphSchema,
+        temporary_directory: Path,
+    ) -> None:
+        # TODO: Fix circular imports
+        from rasa.engine import graph
+
+        for filename, schema in zip(
+            [MODEL_ARCHIVE_TRAIN_SCHEMA_FILE, MODEL_ARCHIVE_PREDICT_SCHEMA_FILE],
+            [train_schema, predict_schema],
+        ):
+            serialized = graph.graph_schema_as_dict(schema)
+            rasa.shared.utils.io.write_yaml(serialized, temporary_directory / filename)
+
+    @staticmethod
+    def _persist_metadata(
+        domain: Domain, model_metadata: Dict[Text, Any], temporary_directory: Path
+    ) -> None:
+        model_metadata = model_metadata.copy()
+        model_metadata[METADATA_DOMAIN_KEY] = domain.as_dict()
+        model_metadata[METADATA_TRAINED_AT_KEY] = time.time()
+        model_metadata[METADATA_MODEL_ID_KEY] = uuid.uuid4().hex
+        model_metadata[METADATA_RASA_VERSION_KEY] = rasa.__version__
+
+        rasa.shared.utils.io.dump_obj_as_json_to_file(
+            temporary_directory / MODEL_ARCHIVE_METADATA_FILE, model_metadata
+        )
+
+    def unpack(
+        self, model_archive_path: Union[Text, Path]
+    ) -> Tuple[GraphSchema, GraphSchema, Domain, ModelMetadata]:
+        """Unpacks a model archive and initializes `ModelStorage`.
```

Comment:
Mmhm, then it would return even more in the tuple 🙈 I think this makes sense to clearly separate the model storage from the additional unpacked stuff but I don't have a strong opinion on it. What do you think?

Reply:
I feel like we should have `create_for` and `create_from` class methods again. I find the `_initialize_model_storage_from_model_archive` stuff a bit weird.

Reply:
They `create_for` and `create_from` ones from the impl proposal were weird as they had different internal state depending on how they were initialized. Happy to go with a classmethod for unpacking though

Reply:
done that